Source/bmalloc/bmalloc/MARRegistry.h

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// MAR: Malloc Audit Records
+//
+// MAR provides a new way to audit bmalloc/libpas memory allocations
+// without resorting to PGM's guard pages. MAR maintains the address
+// of each allocation, but instead tracks what allocations were made
+// within pages of interest through the stack trace when `malloc` is
+// invoked.
+
+#ifndef MAR_REGISTRY_H
+#define MAR_REGISTRY_H
+
+#include <assert.h>
+#include <execinfo.h>
+#include <string.h>
+#include <stdint.h>
+
+#include "MARCrashReporterReport.h"
+
+const unsigned MARProbability = 8192;
+const unsigned MARTrackedBacktraces = 16384;
+const unsigned MARTrackedAllocations = 16384;
+
+// We'll use an approach similar to hardware FIFO; the queue is empty if head == tail
+// and full if head ^ tail == size
+const unsigned MARAllocationRecordTableFIFOModulus = 2 * MARTrackedAllocations;
+
+const unsigned PageSize = 16384;
+const unsigned PageShift = 14;
+
+struct BacktraceRecord {
+    unsigned m_numFrames;
+    uint32_t m_hash;
+    void* m_backtraceBuffer[BacktraceMaxSize];
+};
+
+struct MemoryActionRecord {
+    void* address;
+    size_t m_allocationSize;
+    unsigned m_backtraceRegistryIndex;
+    uint32_t m_backtraceHash;
+    bool isAllocation;
+};
+
+typedef struct MARRegistry MARRegistry;
+struct MARRegistry {
```

Comment:
Just in general this file has been written in a very C-style way, but I can't find any C files which depend on including this code. Is it an artifact of the code having previously been located in libpas, or somehow anticipating such a need?

Just at first glance a lot of this code feels like it could be simplified with a more C++-style approach -- moving methods into classes, making fields private, etc.

Reply:
If it does need to be C-accessible, I think you could get away with a thinner layer of C code, with the C++ structures living underneath the hood behind opaque pointer types.